### PR TITLE
Add NIST keywrap as a cipher mode

### LINKED ISSUE #2003: Add cipher interface to key wrapping modes
We provide the NIST key wrapping modes (KW, KWP) through a custom interface (`mbedtls_nist_kw_xxx`). A key wrapping mode is an AEAD mode, functionally speaking, so it would make sense to provide an interface to these modes through the generic AEAD functions in the cipher module.

The key wrapping modes would only be supported by the one-shot functions (`mbedtls_cipher_auth_encrypt`, `mbedtls_cipher_auth_decrypt`), not by the multipart functions, because they process the full input at once. This is already the case for CCM (which processes the data in a streaming fashion but can't start encrypting until it knows the length of the plaintext) so we wouldn't be changing the philosophy of that API.

Requested by @jethrogb in https://github.com/ARMmbed/mbedtls/pull/1658#issuecomment-417461217

### PR DESCRIPTION
**Pull request by jack-fortanix**
## Description
Adds the NIST keywrap algorithms as AEAD cipher modes. Closes #2003 see also #1658

I had to make two fixes to the cipher tests; they were not resetting the key in the encryption direction (which it seems does not matter for GCM or CCM, but does for NIST keywrap). And, in the test where data is re-encrypted, the output length is compared with the original ciphertext length rather than the original plaintext length. These are identical for GCM/CCM (due to tag being modeled separately) but not for keywrap due to padding.

My understanding is Fortanix already has a corporate CLA with ARM.

## Status
**READY** and waiting review

## Requires Backporting
New feature so probably no.

## Todos
- [x] Tests
- [ ] Documentation
- [x] Changelog updated

I will update docs and changelog after you have reviewed for acceptance.

### CODE REVIEW

**Comment by jack-fortanix** [comment]
Jenkins build is failing but I cannot see the output. If you let me know what is wrong I will fix it.

**Comment by hanno-arm** [comment]
@jack-fortanix Jenkins failing because of a known timing test issue on FreeBSD only. That's as good as a pass.

**Comment by jack-fortanix** [comment]
@RonEld Thanks for review. I have added a Changelog entry, split PR into two commits one with just the fixes to test and the rest for the keywrap impl, and added more test inputs that I believe cover all cases. If there is some additional situation that these are still not hitting just let me know.

File: tests/suites/test_suite_cipher.kw.data [hunk 1]
```diff
@@ -0,0 +1,135 @@
+KW AES-128 wrap rfc 3394
```
Comment (RonEld) [inline tests/suites/test_suite_cipher.kw.data]: please rename the file from `kw` to `nist_kw` as the module name

File: tests/suites/test_suite_cipher.kw.data [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+KWP AES-128 3 byte
+depends_on:MBEDTLS_AES_C:MBEDTLS_NIST_KW_C
+auth_crypt_tv:MBEDTLS_CIPHER_AES_128_KWP:"00000000000000000000000000000000"::"":"":"DAB4EE2853E1C44C5E553E644143902B":"":"":"000102":0
+
+KWP AES-128 4 byte
+depends_on:MBEDTLS_AES_C:MBEDTLS_NIST_KW_C
+auth_crypt_tv:MBEDTLS_CIPHER_AES_128_KWP:"00000000000000000000000000000000"::"":"":"446C037F831092B147C372616357BF7D":"":"":"00010203":0
+
+KWP AES-128 5 bytes
```
Comment (RonEld) [inline tests/suites/test_suite_cipher.kw.data]: All the other tests use the word `byte` in singular

**Comment by jack-fortanix** [comment]
Hi @RonEld I added a Changelog entry in the bugfix commit, and also added a number of tests taken from the NIST CAVS files. I did not add all of them due to it seemed like a lot of data, instead just added every 3rd test from those files. I can easily tweak the script I used to include more or fewer tests as desired. Anyway it now covers {KW,KWP} with all key sizes. I also addressed your other comments with regards to renaming the test suite and fixing the comment.

**Comment by RonEld** [comment]
@jack-fortanix Thank you for addressing my comments.
Indeed this is a lot of test data.
There is no need to have *all* the test vectors from the NIST CAVS files. The idea is to cover all cases. ( e.g. Different input lengths. Too large lengths. Too short lengths. Different padding lengths. Invalid input length . etc..).
I don't see any tests for the unwrapping functionality though.
Please reduce the test data for same use case ( one "COUNT" for every input length is enough) and:
1. Add negative tests
2. Add unwrapping tests.
Thanks in advance

**Comment by jack-fortanix** [comment]
OK I'll prune things down a bit. Re unwrapping, AIUI this is already being tested (the cipher test code first unwraps/decrypts, then wraps/encrypts to confirm identical outputs). However you are right there are no tests for unwrap failing or invalid lengths. I'll address this shortly.

**Comment by jack-fortanix** [comment]
@RonEld I reduced the test cases as you suggested and added some failing tests, also from CAVS.

File: tests/suites/test_suite_cipher.nist_kw.data [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+KWP AES-128 28 byte input
+depends_on:MBEDTLS_AES_C:MBEDTLS_NIST_KW_C
+auth_crypt_tv:MBEDTLS_CIPHER_AES_128_KWP:"00000000000000000000000000000000"::"":"":"EFD0BC7612331A98F2D68F86E606717197BF57E35114234C675D40E9462ACF00DE7860C0F38677F7":"":"":"000102030405060708090A0B0C0D0E0F101112131415161718191A1B":0
+
+KW AES-128 wrap CAVS 17.4 COUNT 0 PLEN 16
+depends_on:MBEDTLS_AES_C:MBEDTLS_NIST_KW_C
+auth_crypt_tv:MBEDTLS_CIPHER_AES_128_KW:"7575da3a93607cc2bfd8cec7aadfd9a6":"":"":"031f6bd7e61e643df68594816f64caa3f56fabea2548f5fb":"":"":"42136d3c384a3eeac95a066fd28fed3f":0
+
+KW AES-128 wrap CAVS 17.4 WTF PLEN 16
```
Comment (RonEld) [inline tests/suites/test_suite_cipher.nist_kw.data]: nice touch:)
Please change to "FAIL"

**Comment by RonEld** [comment]
@jack-fortanix I have backported the bug fix on your behalf to `mbedtls-2.7` and `mbedtls-2.17` in #2519 and #2520

File: tests/suites/test_suite_cipher.nist_kw.data (hunk 3 above)
Comment (jack-fortanix) [reply tests/suites/test_suite_cipher.nist_kw.data]: Haha oops script failure, will fix.

**Comment by jack-fortanix** [comment]
@yanesca Any chance this will be reviewed soon? We are disappointed this didn't make it into the 2.17 release.

**Comment by yanesca** [comment]
@jack-fortanix Unfortunately I have a quite long review queue. I'll try to prioritise this PR and will ask someone else for a review if I can't. Thank you for your patience.

**Comment by jack-fortanix** [comment]
Addressed @yanesca's comments and rebased against latest `development`.

**Comment by jack-fortanix** [comment]
Travis failures are in DTLS nbio test, I believe unrelated to this PR.

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -130,11 +131,6 @@ if(NOT USE_CRYPTO_SUBMODULE)
     add_test_suite(rsa)
     add_test_suite(xtea)
 endif()
-add_test_suite(debug)
```
Comment (RonEld) [inline tests/CMakeLists.txt]: Why were these tests removed? IS it a rebase issue?

Comment (jack-fortanix) [reply tests/CMakeLists.txt]: Rebase error, sorry about that. Fixed now.

**Comment by yanesca** [comment]
As @RonEld mentioned, we still need backports for the bugfix part before we could merge.

@jack-fortanix Could you please raise PRs targeting the mbedtls-2.7 and mbedtls-2.16 branches with the bugfix commit (1dbc5a257fd0767ca741ec18b951f1d666d3a062)?

**Comment by yanesca** [comment]
@jack-fortanix Please don't mind my previous comment, @RonEld had already created the backport branches. Sorry for the noise.

**Pull request merged**